### PR TITLE
ESQL: Make `Source` `Writeable`

### PR DESCRIPTION
**Pull request by nik9000**
We'd love to move all of the serialization from `PlanNamedTypes` into the the types themselves. But serialization for `Source` is in ESQL proper, not ESQL-core with all of the types that need to be serialized. This moves that serialization. It does so by making an interface in ESQL-core that contains the `query` itself which is required to serialize `Source`. This seems like the simplest way to get started moving serialization. The shim *should* be temporary, but it'll take some time to move things around.

### CODE REVIEW

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/SourceUtils.java [hunk 1]
```diff
@@ -21,73 +19,17 @@ public final class SourceUtils {
...
 
-        String text;
-        if (queryText == null) {
-            text = in.readString();
-        } else {
-            int length = in.readInt();
-            text = sourceText(queryText, line, column, length);
-        }
+        String text = in.readString();
         return new Source(new Location(line, charPositionInLine), text);
```
Comment (nik9000) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/SourceUtils.java:33]: I've kept this second serialization mechanism in place for now. It should probably be reworked so we don't use it any more, but that's going to take some more time.

**Review by costin — APPROVED** [review]
LGTM

P.S. I like PlanNamedTypes as it's self contained and helps reason about the system in one place. However for extensibility reasons we'd have to go the other way.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java [hunk 2]
```diff
@@ -25,6 +33,31 @@ public Source(Location location, String text) {
         this.text = text;
     }
 
+    public static Source readFrom(StreamInput in) throws IOException {
```
Comment (alex-spies) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java:36]: I think the method signature should state more clearly that this doesn't work with a general `StreamInput`.
```suggestion
    public static <S extends StreamInput & PlanStreamInput> Source readFrom(S in) throws IOException {
```
Then we also don't need to risk the `ClassCastException` below.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/PlanStreamInput.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * and just we'll remove this when we flatten everything from ESQL-core into
+ * ESQL proper.
+ */
+public interface PlanStreamInput {
+    /**
+     * The query sent by the user to build this plan. This is used to rebuild
+     * {@link Source} without sending the query over the wire over and over
+     * and over again.
+     */
+    String query();
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/PlanStreamInput.java:24]: That's probably better, yeah.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/SourceUtils.java [hunk 4]
```diff
@@ -21,73 +19,17 @@ public final class SourceUtils {
     private SourceUtils() {}
 
     public static void writeSource(StreamOutput out, Source source) throws IOException {
```
Comment (alex-spies) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/SourceUtils.java:21]: Maybe let's mark this as deprecated (for removal)?

Similarly for `readSource`.

**Review by alex-spies — APPROVED** [review]
This is very nice. Just a minor remark.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java (hunk 2 above)
Comment (ChrisHegarty) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java:36]: PlanStreamInput is a StreamInput. So the type could be a PlanStreamInput, no? Or does that force the caller to do the cast ?

**Review by ChrisHegarty — APPROVED** [review]

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java:36]: I'll give this a go. My instincts are that things like this in java don't work well.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/SourceUtils.java (hunk 4 above)
Comment (nik9000) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/SourceUtils.java:21]: Looks like there were two copies of `SourceUtils` somehow. One I can just remove. The other I'll deprecate.

Comment (luigidellaquila) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/SourceUtils.java:27]: Oh yeah, the `SourceUtils` in QL was specific for ESQL, we don't need it anymore

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java]: There's an abstract superclass vs interface thing in the way. It's hopefully temporary, if confusing.

**Pull request merged**